### PR TITLE
Pull request module 5

### PR DESCRIPTION
**Pull request by alvaro-o**
This pull request contains a script to select the best performing stocks out of a dataset with financial data. The dataset includes quarterly data for a few thousands of tickers starting from the year 2006. These data are split into train and test sets in a way to avoid data leakage. A gradient boosting machine classifier is used to determine which stocks outperform SP500 performance in a year from the same date of execution.

Learning curves are plotted to ensure that the model is learning correctly. When overfitting is detected, some modifications are applied to simplify the model. After this, feature importances are examined and plotted with SHAP to check for information leakage. The final results are compared to a baseline to see if the model adds any value, looking at mean and median return across periods.

### CODE REVIEW

File: src/module_5/module5.md [hunk 1]
```diff
@@ -0,0 +1,1005 @@
...
+
+
+```python
+from scipy.stats import lognorm
+import matplotlib.pyplot as plt
+```
+
+BASELINE
+
+First, we start by defining a baseline to which compare the results of the model. In this case, the baseline is the performance of the best 100 stocks last year.
```
Comment (pabaldonedo) [inline src/module_5/module5.md:454]: Quite a strong baseline! SP 500 has 500 stocks. So the baseline is picking the top 20% already.

File: src/module_5/module5.md [hunk 2]
```diff
@@ -0,0 +1,1005 @@
...
+
+
+```python
+def get_baseline(test_results_final_tree):
+    top100_last_period = data_set[data_set.execution_date == '2020-03-31'].sort_values(by='diff_ch_sp500', ascending=False)[["diff_ch_sp500",'Ticker']].head(500)
```
Comment (pabaldonedo) [inline src/module_5/module5.md:459]: head(500) but you wanted to pick top 100?

Comment (pabaldonedo) [inline src/module_5/module5.md:459]: Executiion date is fixed to a date, but it should be changing similarly as we do the training / validation, with the moving window. Otherwise the baseline is just this stocks frozen over time

File: src/module_5/module5.md [hunk 3]
```diff
@@ -0,0 +1,1005 @@
...
+To confirm the algorithm is actually learning correctly, the learning curves can be plotted. As there are results for every execution date, the learning curves can be represented with a set of boxplots instead of points. To see the progress in learning, these boxplots can be compared to each other by setting the first one to zero and normalizing the rest.
+
+Since the current objective is to diagnose the problems within the model, a low level metric such as binary logloss is chosen, leaving 'weighted return' to present the results in a latter moment.
+
+
+```python
+def learning_curves(set_):
+    learning_curves = pd.DataFrame()
+    for date in all_results:
+        learning_curve = pd.DataFrame(all_results[(date)][set_])
```
Comment (pabaldonedo) [inline src/module_5/module5.md:516]: style comment: unnecessary brackets around date

File: src/module_5/module5.md [hunk 4]
```diff
@@ -0,0 +1,1005 @@
...
+
+```
+
+
+    
+![png](module5_files/module5_45_0.png)
+    
+
+
+It seems that the model is overfitting since the test error does not decrease like the training error does as the model gets more complex. To try to fix this, the model can be simplified by modifying some parameters like the number of trees and the learning rate.
```
Comment (pabaldonedo) [inline src/module_5/module5.md:565]: it actually increases!

File: src/module_5/module5.md [hunk 5]
```diff
@@ -0,0 +1,1005 @@
...
+
+```
+
+
+    
+![png](module5_files/module5_49_0.png)
+    
+
+
+By reducing the learning rate, it appears that the test error decreases, which means the fit is better. The median of the error seems to reach its minimum at 16 trees, at which point it starts to overfit again.
```
Comment (pabaldonedo) [inline src/module_5/module5.md:665]: you should replot the training as well to double check we are learning here too

File: src/module_5/module5.md [hunk 6]
```diff
@@ -0,0 +1,1005 @@
...
+
+def train_all_dates():
+    global all_results
+    global all_predicted_tickers_list
+    global all_models 
+    global all_importances
+
+    for execution_date in execution_dates:
+        print(execution_date)
+        all_results,all_predicted_tickers_list,all_models,model,X_train,X_test,all_importances = run_model_for_execution_date(execution_date,all_results,all_predicted_tickers_list,all_models,all_importances,n_trees,learning_rate,False, compute_importances)
```
Comment (pabaldonedo) [inline src/module_5/module5.md:312]: n_trees and learning_rate are neither global nor variables that are received on call. This should not be like this

File: src/module_5/module5.md (hunk 5 above)
Comment (pabaldonedo) [reply src/module_5/module5.md:665]: Also, it is interesting that your results are worse than the solution presented in class. It would be interesting to check:
 - If it is due to path_smooth parameter
 - If it is due to a lot of variance of the model due to different seed.

File: src/module_5/module5.md [hunk 7]
```diff
@@ -0,0 +1,1005 @@
...
+shap.waterfall_plot(all_importances[np.datetime64('2015-06-30T00:00:00.000000000')][0])
+```
+
+
+    
+![png](module5_files/module5_56_0.png)
+    
+
+
+There are some features with a lot of relevance that seem suspicious. For example, "close_0" has an extremely big impact on predictions compared to other features. When looking into it, it results that close_0 leaks information from the future, as the closing price in the data for tickers with counter-splitted stocks was calculated using the current price as a factor.
```
Comment (pabaldonedo) [inline src/module_5/module5.md:786]: Generally speaking, try to justify this kind of statements. Why "close_0" being very important is suspicious? Having a very relevant feature could be reasonable (although unlikely) depending on the context.

**Review by pabaldonedo — APPROVED** [review]
Good in general, just left a handful of comments. I want to highlight these two, though since I find them relevant:

- "You should replot the training as well to double check we are learning here too"
- "

Also, it is interesting that your results are worse than the solution presented in class. It would be interesting to check:

    If it is due to path_smooth parameter
    If it is due to a lot of variance of the model due to different seed.
"


It would have also been interesting for you to explore further directions like those in the discussion section of the solution class
